### PR TITLE
For SG-7259, adds JSON serialization utilities and Context to dict serializer

### PR DESCRIPTION
**Pull request by jfboismenu**
Introduces two features related to the new publisher API:
- sgtk.util.json.{load,loads}, which are thin wrappers around the Python API and allows to json documents into dictionaries using utf-8 encoded strings.
- sgtk.Context.{to,from}_dict, which allow to serialize/unserialize a context to a dictionary.

### CODE REVIEW

File: python/tank/util/json.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    """
+    Converts any :class:`unicode` instances in the input value into a utf-8
+    encoded :class`str` instance.
+
+    :param input_value: Value to convert. Can be a scalar, list or dictionary.
+
+    :returns: A value with utf-8 encoded :class:`str` instances.
+    """
+
+    if isinstance(input_value, unicode):
```
[the remark this answers is not in the document]
Comment (jfboismenu) [reply python/tank/util/json.py]: Yo dawg, I know you are running on Python 3, but unicode is a valid keyword in Python 2...

File: tests/core_tests/test_context.py [hunk 2]
```diff
@@ -1140,6 +1155,25 @@ def test_serialize_without_user(self):
...
+        """
+        ctx = context.Context(**self.kws)
+        ctx_dict = ctx.to_dict()
+        new_ctx = tank.Context.from_dict(ctx.sgtk, ctx_dict)
+        self._assert_equal_contexts(new_ctx, ctx)
+
+    def _assert_equal_contexts(self, ctx_1, ctx_2):
+        """
+        Ensures two contexts are equal.
+        Note that source_entity is not part of the equality comparison,
```
Comment (jfboismenu) [inline tests/core_tests/test_context.py]: @thebeeland Do you remember why that is not the case? It should probably be documented in the __eq__ method.

File: python/tank/context.py [hunk 3]
```diff
@@ -761,12 +754,73 @@ def deserialize(cls, context_str):
...
         tk = Tank(pipeline_config_path)
+        data["tk"] = tk
 
         # add it to the constructor instance
+        # and lastly make the obejct
+        return cls._from_dict(data)
+
+    def to_dict(self):
+        """
+        Converts the context into a dictionary with keys ``project``,
```
Comment (jfboismenu) [inline python/tank/context.py]: I've chosen to not repeat the mistake of serializing the user inside the dictionary, as this has been error prone for clients submitting jobs on their farm where they serialize the context. This pattern feels right.

File: python/tank/context.py [hunk 4]
```diff
@@ -1531,22 +1578,19 @@ def context_yaml_constructor(loader, node):
     """
     # lazy load this to avoid cyclic dependencies
     from .api import Tank
-    
+
     # get the dict from yaml
-    context_constructor_dict = loader.construct_mapping(node)
-    
+    context_constructor_dict = loader.construct_mapping(node, deep=True)
```
Comment (jfboismenu) [inline python/tank/context.py]: Yaml Serialization/unserialization was half-broken. I discovered it when I added the additional_entities value to the Context serialization unit test.

Now it works.

File: python/tank/util/json.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    if isinstance(input_value, dict):
+        return dict(
+            (_convert_unicode_keys_to_string(k), _convert_unicode_keys_to_string(v))
+            for k, v in input_value.iteritems()
+        )
+
+    return input_value
+
+
+def load(
```
Comment (jfboismenu) [inline python/tank/util/json.py]: I've reused the name of the methods in the Python module. Should I have gone for a more verbose name instead, like
load_from_file and load_from_string?

File: python/tank/context.py [hunk 6]
```diff
@@ -761,12 +754,73 @@ def deserialize(cls, context_str):
...
+
+        This ensures that if a more recent version of Toolkit serializes a context
+        and this API reads it that it won't blow-up.
+
+        :param dict data: Data for the context.
+
+        :returns: :class:`Context`
+        """
+        # Get all argument names except for self.
+        argument_names = inspect.getargspec(Context.__init__).args[1:]
```
Comment (manneohrstrom) [inline python/tank/context.py]: this feels quite magical. the `to_dict` is very explicit - any reason not to do the same here? I think that may be easier code to maintain.

**Pull request merged**